src/rpc/blockchain.cpp

```diff
@@ -2140,7 +2142,11 @@ UniValue scantxoutset(const JSONRPCRequest& request)
                 if (!desc->Expand(i, provider, scripts, provider)) {
                     throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("Cannot derive script without private keys: '%s'", desc_str));
                 }
-                needles.insert(scripts.begin(), scripts.end());
+                for (const auto script : scripts) {
+                    std::string desc = InferDescriptor(script, provider)->ToString();
```

Comment:
In commit "Add matching descriptors to scantxoutset output + tests" (fe9b68a81a6f3a9abac7b3b7740d5019129d4fff)

I think I agree with @Sjors in https://github.com/bitcoin/bitcoin/pull/14477#issuecomment-434280868 that behavior that comes from inferring a descriptor here, instead of deriving it directly from `desc` above, is a little unintuitive. IMO, it'd be nicer if `Expand` took an optional `std::vector<std::unique_ptr<Descriptor>>* output_descriptors` argument and filled it in parallel with the existing `output_scripts` argument. 

This is just a suggestion, though, and maybe something to consider for a future PR.

Reply:
That makes sense. The inferred descriptor is functionally equivalent with a more straightforward specialization of the descriptor you're suggesting, but descriptors are intended to at least be somewhat human readable, so this may matter.

However, it's quite a bit of code to implement that (it needs specialization code for a number of constructions, including `combo()` which would need to get split out into ``pkh()` and `pk()` etc), so I'd prefer to keep it for a later PR.

For that reason I also prefer not making any explicit promises about what is put in the RPC outputs; for now all it guarantees is something that encapsulates all the information. For `scantxoutset` we could indeed be better and maintain more of the input descriptor's structure.